src/test/java/com/onaio/steps/handler/actions/TakeSurveyHandlerTest.java

```diff
@@ -161,47 +187,98 @@ public void ShouldNotBeAbleToHandleResultForOtherRequestCode(){
+        surveyHandler.handleResult(new Intent().setData(mock(Uri.class)), Activity.RESULT_OK);
 
         Mockito.verify(householdMock).setStatus(InterviewStatus.INCOMPLETE);
-        Mockito.verify(householdMock).update(Mockito.any(DatabaseHelper.class));
+        Mockito.verify(householdMock).update(any(DatabaseHelper.class));
     }
 
     private void mockSavedForm(ArrayList<IForm> savedForms, String formStatus) {
-        ODKSavedForm savedForm = Mockito.mock(ODKSavedForm.class);
+        ODKSavedForm savedForm = mock(ODKSavedForm.class);
         savedForms.add(savedForm);
-        Mockito.when(savedForm.getStatus()).thenReturn(formStatus);
+        when(savedForm.getStatus()).thenReturn(formStatus);
     }
 
     @Test
     public void ShouldNotHandleResultForErrorResult(){
         takeSurveyHandler.handleResult(null, Activity.RESULT_CANCELED);
 
         Mockito.verify(householdMock,Mockito.never()).setStatus(InterviewStatus.DONE);
-        Mockito.verify(householdMock,Mockito.never()).update(Mockito.any(DatabaseHelper.class));
+        Mockito.verify(householdMock,Mockito.never()).update(any(DatabaseHelper.class));
+    }
+
+    @Test
+    public void testOpenShouldCallStrategyOpenMethodAndHandleException() throws IOException {
+        doNothing().when(takeSurveyForHouseholdStrategy).open();
+        takeSurveyHandler.open();
+
+        verify(takeSurveyForHouseholdStrategy, times(1)).open();
+
+        // handle exception
+        doThrow(NullPointerException.class).when(takeSurveyForHouseholdStrategy).open();
+        takeSurveyHandler.open();
+
+        ArgumentCaptor<Exception> arException = ArgumentCaptor.forClass(Exception.class);
+        verify(exceptionHandler, times(1)).handle(arException.capture());
+
+        assertEquals(NullPointerException.class.getName(), arException.getValue().getClass().getName());
+    }
+
+    @Test
+    public void testGetSavedFormsShouldReturnSavedFormAndHandleException() throws RemoteException {
+        Faker.findODKSavedForm(householdActivitySpy);
+
+        Intent intent = mock(Intent.class);
+        Uri uri = mock(Uri.class);
+
+        when(intent.getData()).thenReturn(uri);
+        when(uri.getLastPathSegment()).thenReturn("");
+
+        List<IForm> savedForms = takeSurveyHandler.getSavedForms(intent);
+        assertFalse(savedForms.isEmpty());
+
+        ODKSavedForm savedForm = (ODKSavedForm) savedForms.get(0);
+        assertEquals("id", savedForm.getId());
+        assertEquals("jrFormId", savedForm.getJrFormId());
+        assertEquals("displayName", savedForm.getDisplayName());
+        assertEquals("complete", savedForm.getStatus());
+
+        // handle exception
+        ContentResolver contentResolver = householdActivitySpy.getContentResolver();
+        when(contentResolver.acquireContentProviderClient(any(Uri.class))).thenReturn(null);
+
+        assertNull(takeSurveyHandler.getSavedForms(intent));
```

Comment:
This should be a separate test on it's own

Reply:
done